File: Solutions/941-Valid Mountain Array/941-Valid Mountain Array.go
//Third fastest
// 模拟就行了。
// 二分不太合适，因为不一定是合法的。
func validMountainArray(arr []int) bool {
    n := len(arr)
    if n < 3 {
        return false
    }
    last := -1
    down := false
    pivot := 0
    for i := 0; i < n; i++ {
        if arr[i] == last {
            return false
        }
        if !down {
            if arr[i] < last {
                down = true
                pivot = i - 1
            }
        } else {
            if arr[i] > last {
                return false
            }
        }
        last = arr[i]
    }
    if !down || pivot==0 {
        return false
    }
    return true
}

//Second fastest
func validMountainArray(arr []int) bool {
    if len(arr) < 3     {return false}
    if arr[1] <= arr[0] {return false}
    if arr[len(arr)-1] >= arr[len(arr)-2] {return false}
    f := true
    for i := 2; i < len(arr)-1; i++ {
        d := arr[i] - arr[i-1]
        if d == 0 { return false }
        if d > 0 {
            if f {continue}
            return false
        } else {
            if !f {continue}
            f = false
        }
    }
    return true
}

//Fastest
func validMountainArray(arr []int) bool {
    // walk up then walk down
    distance := len(arr) -1
    step := 0
    
    for step < distance && canStepUp(arr,step) {
        step++
    }
    
    if step == distance || step == 0 {
        return false
    }
    
    for step < distance && canStepDown(arr,step){
        step++
    }
    
    if step == distance {
        return true
    }
    
    return false
    
}

func canStepUp(arr []int,step int) bool {
    return arr[step] < arr[step+1]
}

func canStepDown(arr []int,step int) bool {
    return arr[step] > arr[step+1]
}

File: Solutions/941-Valid Mountain Array/941-Valid Mountain Array.js
//Third fastest
var validMountainArray = function(arr) {
    if(arr.length < 3)
        return false;
    let i = 0, j = arr.length-1;
    while(i < arr.length - 2 && arr[i] < arr[i+1]){
        i++;
    }
    while(j > 0 && arr[j] < arr[j - 1]){
        j--;
    }
    console.log(`i ${i} j ${j}`)
    if(i === arr.length - 1 || j === 0)
        return false;
    return i == j;
};

//Second fastest
var validMountainArray = function(arr) {
    if(arr.length < 3) return false;
    
    let peak = 0;
    let less = 0;
    let isUp = true;
    let isPyr = true

    for(i=1;i<arr.length; i++) {
        if(arr[i] > arr[peak] && isUp) {
           peak = i
           } else if(arr[i] < arr[peak]) {
               if(isUp) {
                   isUp = false  
                   less = i
               } else if(less && arr[i] < arr[less]) {
                less = i
               isUp = false   
               } else {
                   isPyr = false
               }
          } else {
              isUp = false
              isPyr = false
          }
    }
    
    
    if(peak===0 || peak=== (arr.length -1)) {
        isPyr = false
    }
    
    return isPyr;
};

//Fastest
const validMountainArray = (arr) => {

    //default
    if(arr.length < 3) return false;

    //array is not initially increasing
    if(arr[0] >= arr[1]) return false; 

    //if this point is reached, array is increaing
    let up = true;

    //loop through array 
    for(let i = 1; i< arr.length -1; i++){
        //going down now
        if(arr[i+1] < arr[i]){
            up = false;
        }
        //going back up, so not mountain array
        if(!up && (arr[i+1] >= arr[i])){
            return false;
        }
    }
    //not mountain becuause never decreased
    if(up) return false;
    return true;
}